workspaces/arborist/lib/arborist/build-ideal-tree.js

```diff
@@ -1226,9 +1226,22 @@ This is a one-time fix-up, please be patient...
     const { installLinks, legacyPeerDeps } = this
     const isWorkspace = this.idealTree.workspaces && this.idealTree.workspaces.has(spec.name)
 
-    // spec is a directory, link it unless installLinks is set or it's a workspace
+    // spec is a directory, link it if:
+    // - it's a workspace, OR
+    // - it's a project-internal file: dependency (always linked), OR
+    // - it's external and installLinks is false
     // TODO post arborist refactor, will need to check for installStrategy=linked
-    if (spec.type === 'directory' && (isWorkspace || !installLinks)) {
+    let isProjectInternalFileSpec = false
+    if (edge && edge.rawSpec && edge.rawSpec.startsWith('file:') &&
+        (edge.rawSpec.startsWith('file:../') || edge.rawSpec.startsWith('file:./'))) {
```

Comment:
```suggestion
    if (edge?.rawSpec.startsWith('file:../') || edge?.rawSpec.startsWith('file:./')) {
```

I don't *think* we need `rawSpec?.` I think if it exists it's a string thanks to npm-package-arg.